SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestHandoverReplication/SpatialTestHandoverReplication.cpp

```diff
@@ -0,0 +1,213 @@
+﻿// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "SpatialTestHandoverReplication.h"
+#include "EngineClasses/SpatialNetDriver.h"
+#include "Kismet/GameplayStatics.h"
+#include "LoadBalancing/LayeredLBStrategy.h"
+
+#include "DynamicReplicationHandoverCube.h"
+#include "SpatialFunctionalTestFlowController.h"
+
+/**
+ * This tests that an Actor's bReplicate flag is properly handed over when dynamically set.
+ * Tests UNR-4441
+ * This test contains 4 Server and 2 Client workers.
+ *
+ * The flow is as follows:
+ * - Setup:
+ *	- Server 1 spawns a HandoverCube (called ADynamicReplicationHandoverCube) with bReplicates set to false inside its authority area.
+ *	- The bReplicates flag is set to true after the end of the Actor's initialization.
+ *  - All servers set a reference to the HandoverCube and reset their local copy of the LocationIndex and the AuthorityCheckIndex.
+ * - Test:
+ *	- At this stage, Server 1 should have authority over the HandoverCube.
+ *  - The HandoverCube moves into the authority area of Server 2.
+ *  - At this stage, Server 2 should have authority over the HandoverCube.
+ *  - Server 2 acquires a lock on the HandoverCube and moves it into the authority area of Server 3.
+ *	- Since Server 2 has the lock on the HandoverCube it should still be authoritative over it.
+ *  - Server 2 releases the lock on the HandoverCube.
+ *  - At this point, Server 3 should become authoritative over the HandoverCube.
+ *  - The HandoverCube moves into the authority area of Server 4.
+ *  - At this point, Server 4 should be authoritative over the Handover	Cube.
+ * - Clean-up:
+ *	- The HandoverCube is destroyed.
+ */
+
+ASpatialTestHandoverReplication::ASpatialTestHandoverReplication()
+	: Super()
+{
+	Author = "Antoine Cordelle";
+	Description = TEXT("Test dynamically set replication for an actor");
+
+	Server1Position = FVector(-500.0f, -500.0f, 50.0f);
+	Server2Position = FVector(500.0f, -500.0f, 50.0f);
+	Server3Position = FVector(-500.0f, 500.0f, 50.0f);
+	Server4Position = FVector(500.0f, 500.0f, 50.0f);
+}
+
+void ASpatialTestHandoverReplication::PrepareTest()
+{
+	Super::PrepareTest();
+
+	AddStep(TEXT(
+		"Server 1 spawns a HandoverCube (called ADynamicReplicationHandoverCube) with bReplicates set to false inside its authority area."), FWorkerDefinition::Server(
+		        1), nullptr, [this]()
+	        {
+		        HandoverCube =
+			        GetWorld()->SpawnActor<ADynamicReplicationHandoverCube>(Server1Position, FRotator::ZeroRotator, FActorSpawnParameters());
+		        RegisterAutoDestroyActor(HandoverCube);
+		        FinishStep();
+	        });
+
+	AddStep(TEXT("Server sets Actor's replication to true"), FWorkerDefinition::Server(1), nullptr, [this]() {
+		HandoverCube->SetReplicates(true);
```

Comment:
What's the reasoning behind doing this in a separate step?

Reply:
This is the central thing that is being tested, so I put it in a separate step to highlight this. This is not necessary code-wise though.